src/ephemeral.rs

```diff
@@ -396,72 +398,58 @@ WHERE
-            r#"
-    SELECT ephemeral_timestamp
-    FROM msgs
-    WHERE ephemeral_timestamp != 0
-      AND chat_id != ?
-    ORDER BY ephemeral_timestamp ASC
-    LIMIT 1;
-    "#,
-            paramsv![DC_CHAT_ID_TRASH], // Trash contains already deleted messages, skip them
-        )
-        .await
-    {
-        Err(err) => {
-            warn!(context, "Can't calculate next ephemeral timeout: {}", err);
-            return;
-        }
-        Ok(ephemeral_timestamp) => ephemeral_timestamp,
-    };
+    Ok(())
+}
 
-    // Cancel existing task, if any
-    if let Some(ephemeral_task) = context.ephemeral_task.write().await.take() {
-        ephemeral_task.cancel().await;
-    }
+pub(crate) async fn ephemeral_loop(context: &Context, interrupt_receiver: Receiver<()>) {
+    loop {
+        let ephemeral_timestamp: Option<i64> = match context
+            .sql
+            .query_get_value(
+                r#"
+                SELECT min(ephemeral_timestamp)
+                FROM msgs
+                WHERE ephemeral_timestamp != 0
+                  AND chat_id != ?;
+                "#,
+                paramsv![DC_CHAT_ID_TRASH], // Trash contains already deleted messages, skip them
+            )
+            .await
+        {
+            Err(err) => {
+                warn!(context, "Can't calculate next ephemeral timeout: {}", err);
+                None
+            }
+            Ok(ephemeral_timestamp) => ephemeral_timestamp,
+        };
 
-    if let Some(ephemeral_timestamp) = ephemeral_timestamp {
         let now = SystemTime::now();
-        let until = UNIX_EPOCH
-            + Duration::from_secs(ephemeral_timestamp.try_into().unwrap_or(u64::MAX))
-            + Duration::from_secs(1);
+        let until = if let Some(ephemeral_timestamp) = ephemeral_timestamp {
+            UNIX_EPOCH
+                + Duration::from_secs(ephemeral_timestamp.try_into().unwrap_or(u64::MAX))
+                + Duration::from_secs(1)
+        } else {
+            // no messages to be deleted for now, wait long for one to occur
+            now + Duration::from_secs(86400)
+        };
 
         if let Ok(duration) = until.duration_since(now) {
-            // Schedule a task, ephemeral_timestamp is in the future
-            let context1 = context.clone();
-            let ephemeral_task = task::spawn(async move {
-                async_std::task::sleep(duration).await;
-                context1.emit_event(EventType::MsgsChanged {
-                    chat_id: ChatId::new(0),
-                    msg_id: MsgId::new(0),
-                });
-            });
-            *context.ephemeral_task.write().await = Some(ephemeral_task);
-        } else {
-            // Emit event immediately
-            context.emit_event(EventType::MsgsChanged {
-                chat_id: ChatId::new(0),
-                msg_id: MsgId::new(0),
-            });
+            if timeout(duration, interrupt_receiver.recv()).await.is_ok() {
```

Comment:
Maybe add an `info!` log right above saying that the loop has started waiting for interrupt.

Reply:
Like this?